pkg/noderesourcetopology/cache/overreserve.go

```diff
@@ -0,0 +1,228 @@
+// 2. it was pessimistically overallocated, so the node is a candidate for resync
+// This function enables the caller to know the slice of nodes should be considered for resync,
+// avoiding the need to rescan the full node list.
+func (ov *OverReserve) DirtyNodeNames() []string {
+	ov.lock.Lock()
+	defer ov.lock.Unlock()
+	var nodes []string
+	for node := range ov.assumedResources {
+		if !ov.nodeDiscarded.IsSet(node) {
+			// noone asked about this node, let it be
+			continue
+		}
+		nodes = append(nodes, node)
+	}
+	return nodes
+}
+
+// Resync implements the cache resync loop step. This function checks if the latest available NRT information received matches the
+// state of a dirty node, for all the dirty nodes. If this is the case, the cache of a node can be Flush()ed.
+// The trigger for attempting to resync a node is not just that we overallocated it. If a node was overallocated but still has capacity,
+// we keep using it. But we cannot predict when the capacity is too low, because that would mean predicting the future workload requests.
+// The best heuristic found so far is count how many times the node was skipped *AND* crosscheck with its overallocation state.
+// If *both* a node has pessimistic overallocation accounted to it *and* was discarded "too many" (how much is too much is a runtime parameter
+// which needs to be set and tuned) times, then it becomes a candidate for resync. Just using one of these two factors would lead to
+// too aggressive resync attempts, so to more, likely unnecessary, computation work on the scheduler side.
+func (ov *OverReserve) Resync() {
+	nodeNames := ov.DirtyNodeNames()
+	// avoid as much as we can unnecessary work and logs.
+	if len(nodeNames) == 0 {
+		klog.V(6).InfoS("nrtcache: resync: no dirty nodes detected")
+		return
+	}
+
+	// we are not working with a specific pod, so we need a unique key to track this flow
+	logID := logIDFromTime()
+	klog.V(6).InfoS("nrtcache: resync NodeTopology cache starting", "logID", logID)
+	defer klog.V(6).InfoS("nrtcache: resync NodeTopology cache complete", "logID", logID)
+
+	var nrtUpdates []*topologyv1alpha1.NodeResourceTopology
+	for _, nodeName := range nodeNames {
+		nrtCandidate, err := ov.nrtLister.Get(nodeName)
+		if err != nil {
+			klog.V(3).InfoS("nrtcache: failed to get NodeTopology", "logID", logID, "node", nodeName, "error", err)
+			continue
+		}
+		if nrtCandidate == nil {
+			klog.V(3).InfoS("nrtcache: missing NodeTopology", "logID", logID, "node", nodeName)
+			continue
+		}
+
+		pfpExpected := podFingerprintForNodeTopology(nrtCandidate)
+		if pfpExpected == "" {
+			klog.V(3).InfoS("nrtcache: missing NodeTopology podset fingerprint data", "logID", logID, "node", nodeName)
+			continue
+		}
+
+		klog.V(6).InfoS("nrtcache: trying to resync NodeTopology", "logID", logID, "node", nodeName, "fingerprint", pfpExpected)
+
+		err = checkPodFingerprintForNode(logID, ov.nodeIndexer, nodeName, pfpExpected)
```

Comment:
(move discussion of https://github.com/kubernetes-sigs/scheduler-plugins/pull/437#discussion_r1021133275 here.)

Let me put it another way: suppose a node has pod1, pod2 scheduled, and NRT has properly get populated with the pfp signature annotation info. Now pod3 comes in and gets (over-)reserved (maybe bound or maybe not), and pod4 comes in but not fit in the node, so it triggers `Resync()`.

At this moment, the combination of pfpExpected and the pods ov.nodeIndexer can be:

| pfpExpected        | ov.nodeIndexer     | result                                                                 |
|--------------------|--------------------|------------------------------------------------------------------------|
| [pod1, pod2]       | [pod1, pod2]       | match, flush the node incorrectly<br>b/c over-reserved pod3 is flushed |
| [pod1, pod2]       | [pod1, pod2, pod3] | mismatch                                                               |
| [pod1, pod2, pod3] | [pod1, pod2, pod3] | match, flush the node correctly                                        |
| [pod1, pod2, pod3] | [pod1, pod2]       | (this might be possible if pod3 is deleted)<br>mismatch                |

So in the first row's case, pod4's next scheduling result could be incorrect b/c over-reserved pod3 has been flushed. This case can be simulated by plumbing a time.Sleep() in binding cycle. Let me know if I misunderstand any logic here.

Reply:
Thanks for raising this point. I think you are right. If we trigger a resync:
1. AFTER a pod was overreserved
2. BUT BEFORE the pod is running
we can incorrectly flush the internal cache, lose the over-reservation and incorrectly schedule.

The fix could be adding the pods we reserved to the indexer immediately (or in general tracking the overreserved pods immediately).
The indexer should handle transparently deletion, so this fix seems pretty self-contained and safe.

Reply:
Thinking more about it, we should surely track these pods, but just adding them to the nodeIndexer is possibly not the best approach. I'll explore options (e.g. separate tracking map, other options) here.